1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+No changes is needed in tools like [PowerShell EditorSyntax](https://github.com/PowerShell/EditorSyntax).
+
+Try to beat any scenario as part of this resource-centric approach and you will see that the approach covers this scenario even `ForEach-Object -Parallel`) automatically (including dot sourcing).
+(In fact, there can certainly be scenarios where this is impossible in principle, like remoting and serialization, but there may be tradeoffs and workarounds)
+
+### Context / Scope
+
+The `Engine` should take into account Context / Scope.
+
+Since variables can be copied to other scopes, the `Engine` must maintain a _usage counter_ during variable assignment / creation operations.
+This adds a bit of complexity, but it should not affect performance, as in most cases, the `Engine` will only have to check a flag and only in very rare cases do extra work.
+Even an `IDisposable` check needs to be done only once.
+Moreover, we can even make a cache for all known types using our `TypeGen.ps1` utility.
+
+Another feature is that native objects can turn around in `PSObject` and the `Engine` should understand this too.
+
+### Annotating types / objects
+
+There are four options for annotating:
+
+- make a cache for all known types using our `TypeGen.ps1` utility
+- use ETS to enhance some types with this feature with default options
+
+    For example, if we define a TempFile type then users will not have to worry about deleting temporary files created by New-TemporaryFile cmdlet at all
+- embed this in the cmdlets so that they issue an already annotated object into the pipeline
+
+    For example, the same New-TemporaryFile or Open-Database cmdlet could do the annotation internally so that users have no need do this explicitly
+- explicitly annotate like `$tempFile.MakeDisposable(DisposableKind.AtRunspaceClose, { script to remove the temp file })`
+
+Note that there are many APIs that use disposable classes and now developers will have the opportunity to freely write objects of these types in the pipeline without fear of resource leakage.
+
+### Kinds of annotation
+
+Mentioned `DisposableKind` is based on context / scope and could be:
+
+- `Unknown`, it is default for new object. The `Engine` should annoutate the object. The default allow the `Engine` to process old code.
+- `Auto`, the `Engine` annoutates based on current context/scope
+- `AtGlobal`, dispose at default runspace close
+- `AtModuleUnload`
+- `AtRunspaceClose`
+- `AtCmdletClose/AtPipelineClose`
+- `AtScriptBlockClose`, dispose at current script block scope close
+
+Thus, a developer is always able to set a desired behavior.
+
+Moreover, we can always call MakeDisposable() in the script to adjust the behavior if the `Engine` cannot do this automatically.
+In this case, we will also need a method to force the mandatory code to execute like `$tempFile.PSDispose()`.
+It may also be required if we want to free an expensive resource before an end of the script block.
```

Comment:
The implementation being suggested isn't super clear here. Are you suggesting additional magic methods be added for this?

Those tend to have discoverability problems in general, I'm not sure that's an effective way to go about it. 😕

Reply:
RFC shouldn't discuss implementation details until this is needed for understanding the concept. Prototypes exist for this.

Reply:
Given the nature of the RFC, I'm not sure how a useful discussion can come from avoiding discussion of some implementation details, to be honest. If you would like to make a prototype to demonstrate aspects, feel free.

Reply:
I have no plans to dive into it until MSFT team says we would like to have.